File: ta-worker/app/factors.py
"""Pure technical-var math. No I/O. Hand-verifiable."""
from __future__ import annotations

import math

import numpy as np

MIN_BARS_X = 20


class InsufficientBars(Exception):
    """Raised when a factor cannot be computed from the given bar count."""


def compute_x_var(closes: list[float]) -> float:
    """X Var (vol factor): symmetric crowding x magnitude dispersion around
    the fitted regression centerline.

    X = (count_below * avg_dist_below) * (count_above * avg_dist_above)
    Returns 0.0 when all residuals sit on one side (degenerate dispersion).
    """
    n = len(closes)
    if n < MIN_BARS_X:
        raise InsufficientBars(f"need >= {MIN_BARS_X} closes, got {n}")
    x = np.arange(n, dtype=float)
    y = np.asarray(closes, dtype=float)
    if not np.isfinite(y).all():
        raise ValueError("closes contains NaN or Inf values")
    slope, intercept = np.polyfit(x, y, 1)
    resid = y - (slope * x + intercept)
    below = resid[resid < 0]
    above = resid[resid > 0]
    if below.size == 0 or above.size == 0:
        return 0.0
    avg_below = float(np.abs(below).mean())
    avg_above = float(np.abs(above).mean())
    result = (below.size * avg_below) * (above.size * avg_above)
    if not math.isfinite(result):
        raise ValueError("x_var overflow: outlier price data")
    return result


def compute_y_var(price: float, target: float, stop: float) -> float:
    """Y Var (risk/reward): distance-to-target / distance-to-stop.
    Returns 0.0 when risk is non-positive (no valid downside reference)."""
    risk = price - stop
    if risk <= 0:
        return 0.0
    return (target - price) / risk


MIN_BARS_Z = 30


def pivot_highs(highs: list[float], w: int = 5) -> list[tuple[int, float]]:
    """Local maxima: index i where highs[i] == max of the +/- w window.

    Assumes `highs` are raw stored OHLC values compared by value; for
    computed/derived highs a tolerance would be needed (exact float
    equality is unreliable for derived values).
    """
    out: list[tuple[int, float]] = []
    n = len(highs)
    for i in range(w, n - w):
        if highs[i] == max(highs[i - w : i + w + 1]):
            out.append((i, float(highs[i])))
    return out


def compute_z_var(
    highs: list[float],
    lows: list[float],
    closes: list[float],
    atr: float,
    lookback: int = 252,
    tol_atr: float = 0.5,
) -> int:
    """Z Var (momentum): count of resistance levels that later flipped to
    support within the last `lookback` bars.

    A flip = a pivot-high level L that price (a) closed above by > tol band
    (breakout), then (b) returned so a bar low touched within tol of L while
    its close held at/above L (acted as support), with the support touch
    occurring inside the lookback window.
    """
    n = len(closes)
    if n < MIN_BARS_Z:
        raise InsufficientBars(f"need >= {MIN_BARS_Z} bars, got {n}")
    if atr <= 0:
        return 0
    band = tol_atr * atr
    window_start = max(0, n - lookback)
    flips = 0
    seen_levels: set[float] = set()
    for pidx, level in pivot_highs(highs, w=5):
        if pidx >= n - 10:
            continue
        level_key = round(level / band) * band
        if level_key in seen_levels:
            continue
        broke = False
        for j in range(pidx + 1, n):
            if not broke:
                if closes[j] > level + band:
                    broke = True
            else:
                touched_as_support = (
                    lows[j] <= level + band
                    and closes[j] >= level - band
                    and j >= window_start
                )
                if touched_as_support:
                    flips += 1
                    seen_levels.add(level_key)
                    break
    return flips

File: app/lib/ta/factors.ts
// Pure technical-var math. No I/O. Ported 1:1 from ta-worker/app/factors.py.

export class InsufficientBars extends Error {}

export const MIN_BARS_X = 20
export const MIN_BARS_Z = 30

/** Degree-1 least-squares fit. Returns [slope, intercept] for y over x = 0..n-1. */
function linfit(y: readonly number[]): [number, number] {
  const n = y.length
  let sx = 0, sy = 0, sxx = 0, sxy = 0
  for (let i = 0; i < n; i++) {
    sx += i; sy += y[i]; sxx += i * i; sxy += i * y[i]
  }
  const denom = n * sxx - sx * sx
  const slope = denom === 0 ? 0 : (n * sxy - sx * sy) / denom
  const intercept = (sy - slope * sx) / n
  return [slope, intercept]
}

const mean = (xs: readonly number[]): number => xs.reduce((a, b) => a + b, 0) / xs.length

/**
 * X Var (vol factor): symmetric crowding × magnitude dispersion around the
 * fitted regression centerline. Returns 0 when all residuals sit on one side.
 */
export function computeXVar(closes: readonly number[]): number {
  const n = closes.length
  if (n < MIN_BARS_X) throw new InsufficientBars(`need >= ${MIN_BARS_X} closes, got ${n}`)
  if (!closes.every(Number.isFinite)) throw new Error('closes contains NaN or Inf values')
  const [slope, intercept] = linfit(closes)
  const below: number[] = [], above: number[] = []
  for (let i = 0; i < n; i++) {
    const resid = closes[i] - (slope * i + intercept)
    if (resid < 0) below.push(resid)
    else if (resid > 0) above.push(resid)
  }
  if (below.length === 0 || above.length === 0) return 0.0
  const avgBelow = mean(below.map(Math.abs))
  const avgAbove = mean(above.map(Math.abs))
  const result = (below.length * avgBelow) * (above.length * avgAbove)
  if (!Number.isFinite(result)) throw new Error('x_var overflow: outlier price data')
  return result
}

/** Y Var (risk/reward): distance-to-target / distance-to-stop. 0 when risk <= 0. */
export function computeYVar(price: number, target: number, stop: number): number {
  const risk = price - stop
  if (risk <= 0) return 0.0
  return (target - price) / risk
}

/** Local maxima: index i where highs[i] equals the max of the ±w window. */
export function pivotHighs(highs: readonly number[], w = 5): Array<[number, number]> {
  const out: Array<[number, number]> = []
  const n = highs.length
  for (let i = w; i < n - w; i++) {
    let m = -Infinity
    for (let k = i - w; k <= i + w; k++) if (highs[k] > m) m = highs[k]
    if (highs[i] === m) out.push([i, highs[i]])
  }
  return out
}

/**
 * Z Var (momentum): count of resistance levels that later flipped to support
 * within the last `lookback` bars. A flip = a pivot-high level that price
 * closed above by > tol band (breakout), then returned so a bar low touched
 * within tol while its close held at/above the level (acted as support).
 */
export function computeZVar(
  highs: readonly number[],
  lows: readonly number[],
  closes: readonly number[],
  atr: number,
  lookback = 252,
  tolAtr = 0.5,
): number {
  const n = closes.length
  if (n < MIN_BARS_Z) throw new InsufficientBars(`need >= ${MIN_BARS_Z} bars, got ${n}`)
  if (atr <= 0) return 0
  const band = tolAtr * atr
  const windowStart = Math.max(0, n - lookback)
  let flips = 0
  const seenLevels = new Set<number>()
  for (const [pidx, level] of pivotHighs(highs, 5)) {
    if (pidx >= n - 10) continue
    const levelKey = Math.round(level / band) * band
    if (seenLevels.has(levelKey)) continue
    let broke = false
    for (let j = pidx + 1; j < n; j++) {
      if (!broke) {
        if (closes[j] > level + band) broke = true
      } else {
        const touchedAsSupport =
          lows[j] <= level + band && closes[j] >= level - band && j >= windowStart
        if (touchedAsSupport) {
          flips += 1
          seenLevels.add(levelKey)
          break
        }
      }
    }
  }
  return flips
}
